src/changeset.rs

```diff
@@ -0,0 +1,192 @@
+                    .iter()
+                    .filter_map(|validation| validation(attribute).err())
+                    .collect();
+                error_list.append(&mut errors_for_element);
+                error_list
+            },
+        );
+        if errors.len() > 0 {
+            Err(ChangeError::ValidationError(errors_acc))
+        } else {
+            Ok(())
+        }
+    }
+}
+impl<Attr, Data> Insertable for Changeset<Attr, Data>
+where
+    Attr: Iden + Valuable + Clone + 'static + std::fmt::Debug,
+    Data: TableRef,
+{
+    fn to_sql(&self) -> String {
+        let changes: Vec<Attr> = self
+            .changes
+            .iter()
+            .map(|Change(attr, _)| attr.clone())
+            .collect();
+
+        Query::insert()
+            .into_table(Data::table_ref())
+            .columns(changes.clone())
+            .values_panic(changes.clone().iter().map(|x| x.value()))
+            .returning(Query::returning().all())
+            .to_string(PostgresQueryBuilder)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    #[derive(Debug, Clone, PartialEq, Eq, Iden)]
+    enum BogusAttrs {
+        Field(String),
+        OtherField(String),
+    }
+    impl Valuable for BogusAttrs {
+        fn value(&self) -> SimpleExpr {
+            match self {
+                Self::Field(val) => val.into(),
+                Self::OtherField(val) => val.into(),
+            }
+        }
+    }
+
+    #[derive(Debug, sqlx::FromRow)]
+    struct Bogus {}
+    impl TableRef for Bogus {
+        fn table_ref() -> Alias {
+            Alias::new("boguses")
+        }
+    }
+    fn setup() -> Changeset<BogusAttrs, Bogus> {
+        let cs: Changeset<BogusAttrs, Bogus> = Changeset::new(None);
+        cs
+    }
+    mod add_change_tests {
+        use super::*;
+
+        #[test]
+        #[should_panic]
+        fn will_not_accept_duplicate_change_types() {
+            let mut cs = super::setup();
+            cs.add_change(BogusAttrs::Field("".to_string()), None);
+            cs.add_change(BogusAttrs::Field("".to_string()), None);
+        }
+
+        #[test]
+        fn will_accept_distinct_change_types() {
+            let mut cs = super::setup();
+            cs.add_change(BogusAttrs::Field("Whatever".to_string()), None);
+            cs.add_change(BogusAttrs::OtherField("Whatever".to_string()), None);
+            let changes: Vec<BogusAttrs> =
+                cs.changes.iter().map(|change| change.0.clone()).collect();
+
+            assert_eq!(
+                vec![
+                    BogusAttrs::Field("Whatever".to_string()),
+                    BogusAttrs::OtherField("Whatever".to_string())
+                ],
+                changes
+            )
+        }
+    }
+    mod validate_tests {
+        use super::*;
+
+        #[test]
+        fn fails_if_validations_not_passing() {
+            let mut cs = super::setup();
+            let too_long = |given_field: &BogusAttrs| {
```

Comment:
If anyone is ambitious we could rip out this changeset module in favor of https://github.com/Keats/validator or similar. I gave a stab at this, but ran into some snags, so I stopped since the changeset logic was already working.